### PR TITLE
feat: v29

### LINKED ISSUE #11387: Validate regex manager combination strategy
### Describe the proposed change(s).

The `combination` strategy seems buggy to me, as we itereate two times over all matches.
This should only happen once.

https://github.com/renovatebot/renovate/blob/8e3f184c717115b574213e49fd6618a140c85ae0/lib/manager/regex/index.ts#L154-L167

### LINKED ISSUE #12155: Disable inferring credentials from environment variables
(Originally discussed at https://github.com/renovatebot/renovate/discussions/12118)

### What would you like Renovate to be able to do?

I don't want Renovate to guess credentials from my environment variables so that I have full control over which credentials are presented to which registry.

### If you have any ideas on how this should be implemented, please tell us here.

I'd like to have a configuration option to disable this behavior. Maybe `inferCredentialsFromEnv`?

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by rarkins**
Rollup of v29 features, to be rebase merged

Closes #12155, Closes #11944, Closes #11387

### CODE REVIEW

**Comment by HonkingGoose** [comment]
Do we maybe want to include breaking changes for the config into this release as well? Or do we want to save those for a later release, so I'm not holding up things now?

I'm talking specifically about:

- #11987
- #12024

**Comment by rarkins** [comment]
> Do we maybe want to include breaking changes for the config into this release as well? Or do we want to save those for a later release, so I'm not holding up things now?
> 
> I'm talking specifically about:
> 
> * [Add `helpers:pinGitHubActionDigests` to `config:base` #11987](https://github.com/renovatebot/renovate/issues/11987)
> * [Rename `config:base` -> `config:recommended` #12024](https://github.com/renovatebot/renovate/issues/12024)

Let's wait on those. I'm not so sure about the first one either.

File: docs/usage/configuration-options.md [hunk 1]
```diff
@@ -763,7 +763,8 @@ Configuration added here applies for all Go-related updates, however currently t
 
 For self-hosted users, `GOPROXY`, `GONOPROXY` and `GOPRIVATE` environment variables are supported ([reference](https://golang.org/ref/mod#module-proxy)).
 
-But when you use the `direct` or `off` keywords Renovate will fallback to its own fetching strategy (i.e. directly from GitHub, etc).
+Usage of `direct` will fallback to Renovate-native release fetching mechanism.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:766]:
```suggestion
Usage of `direct` will fallback to the Renovate-native release fetching mechanism.
```

File: docs/usage/configuration-options.md [hunk 2]
```diff
@@ -763,7 +763,8 @@ Configuration added here applies for all Go-related updates, however currently t
...
+Also we support `off` keyword which immediately will stop any fetching.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:767]:
```suggestion
Also we support the `off` keyword which will stop any fetching immediately.
```

File: docs/usage/configuration-options.md [hunk 3]
```diff
@@ -1667,7 +1668,7 @@ If enabled Renovate will pin Docker images by means of their SHA256 digest and n
 
 ## platformAutomerge
 
-If you have enabled `automerge` and set `automergeType=pr` in the Renovate config, then `platformAutomerge` is enabled by default to speed up merging via the platform's native automerge functionality.
+If you have enabled `automerge` and set `automergeType=pr` in the Renovate config, then you can also set `platformAutomerge` to true to speed up merging via the platform's native automerge functionality.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1671]:
```suggestion
If you have enabled `automerge` and set `automergeType=pr` in the Renovate config, then you can also set `platformAutomerge` to `true` to speed up merging via the platform's native automerge functionality.
```

File: docs/usage/configuration-options.md [hunk 4]
```diff
@@ -1676,7 +1677,10 @@ Any PR that is being updated will be automerged with the Renovate-based automerg
 This option is available for Azure, GitHub and GitLab.
 It falls back to Renovate-based automerge if the platform-native automerge is not available.
 
-Though this option is enabled by default, you can fine tune the behavior by setting `packageRules` if you want to use it selectively (e.g. per-package).
+You can also fine tune the behavior by setting `packageRules` if you want to use it selectively (e.g. per-package).
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1680]:
```suggestion
You can also fine-tune the behavior by setting `packageRules` if you want to use it selectively (e.g. per-package).
```

File: docs/usage/configuration-options.md [hunk 5]
```diff
@@ -1676,7 +1677,10 @@ Any PR that is being updated will be automerged with the Renovate-based automerg
...
+
+If you rely on Renovate's behavior of rebasing automerging branches when `rebaseWhen=auto`, then note that you are no longer guaranteed of up-to-date branches when automerging if you enable platform automerge, unless you use some type of branch protection rules to achieve the same.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1682]: This sentence is too long, and needs to be split up or rewritten. It's too hard and understand in one go.

File: lib/manager/regex/index.ts [hunk 6]
```diff
@@ -182,8 +157,10 @@ function handleRecursive(
     // if we have a depName and a currentValue with have the minimal viable definition
```
Comment (HonkingGoose) [inline lib/manager/regex/index.ts:157]: `with have the` is not valid grammar, should probably be changed to:

```suggestion
    // if we have a depName and a currentValue which have the minimal viable definition
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by zharinov — DISMISSED** [review]
Seems valid to me from the perspective of my GOPROXY-related PR

**Review by viceice — DISMISSED** [review]
Just the mentioned wording should be done. 🤗

File: lib/workers/global/config/parse/env.spec.ts [hunk 7]
```diff
@@ -158,79 +158,6 @@ describe('workers/global/config/parse/env', () => {
...
-        PYPI_FOO_TOKEN: 'some-token',
-      };
-      expect(env.getConfig(envParam).hostRules).toHaveLength(0);
-    });
-    it('rejects npm env', () => {
-      const envParam: NodeJS.ProcessEnv = {
-        npm_package_devDependencies__types_registry_auth_token: '4.2.0',
-      };
-      expect(env.getConfig(envParam).hostRules).toHaveLength(0);
-    });
```
Comment (zharinov) [inline lib/workers/global/config/parse/env.spec.ts:233]: Is it okay to remove these tests? The feature is still available, isn't it? Why get rid of regression tests?

File: docs/usage/configuration-options.md (hunk 5 above)
Comment (rarkins) [inline docs/usage/configuration-options.md:1682]:
```suggestion
If you rely on Renovate's behavior of rebasing automerging branches when `rebaseWhen=auto`, then note that you are no longer guaranteed of up-to-date branches when automerging if you enable platform automerge.
```

**Review by zharinov — DISMISSED** [review]

File: docs/usage/configuration-options.md [hunk 8]
```diff
@@ -1676,7 +1677,10 @@ Any PR that is being updated will be automerged with the Renovate-based automerg
 This option is available for Azure, GitHub and GitLab.
 It falls back to Renovate-based automerge if the platform-native automerge is not available.
 
-Though this option is enabled by default, you can fine tune the behavior by setting `packageRules` if you want to use it selectively (e.g. per-package).
+You can also fine-tune the behavior by setting `packageRules` if you want to use it selectively (e.g. per-package).
+
+If you rely on Renovate's behavior of rebasing automerging branches when `rebaseWhen=auto`, then note that you are no longer guaranteed of up-to-date branches when automerging if you enable platform automerge.
+For example, GitHub might automerge a Renovate branch even if it is one or more commits behind the base branch at the time.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1683]:
```suggestion
For example, GitHub might automerge a Renovate branch even if it's behind the base branch at the time.
```

I think we can further simplify this text.

File: lib/workers/global/config/parse/env.spec.ts (hunk 7 above)
Comment (viceice) [reply lib/workers/global/config/parse/env.spec.ts:233]: I think they are moved.

**Review by viceice — DISMISSED** [review]

File: docs/usage/configuration-options.md [hunk 9]
```diff
@@ -1676,7 +1677,10 @@ Any PR that is being updated will be automerged with the Renovate-based automerg
 This option is available for Azure, GitHub and GitLab.
 It falls back to Renovate-based automerge if the platform-native automerge is not available.
 
-Though this option is enabled by default, you can fine tune the behavior by setting `packageRules` if you want to use it selectively (e.g. per-package).
+You can also fine-tune the behavior by setting `packageRules` if you want to use it selectively (e.g. per-package).
+
+If you rely on Renovate's behavior of rebasing automerging branches when `rebaseWhen=auto`, then note that you are no longer guaranteed of up-to-date branches when automerging if you enable platform automerge.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1682]:
```suggestion
Note that `rebaseWhen=auto` and `platformAutomerge=true` are not fully compatible.
Normally when you set `rebaseWhen=auto` Renovate rebases any branch that's behind the base branch automatically.
This behavior is no longer guaranteed when you enable `platformAutomerge`.
```

How about something like this?

Not sure if the text matches the code/behavior, so check before applying this suggestion.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md [hunk 10]
```diff
@@ -1676,7 +1677,12 @@ Any PR that is being updated will be automerged with the Renovate-based automerg
 This option is available for Azure, GitHub and GitLab.
 It falls back to Renovate-based automerge if the platform-native automerge is not available.
 
-Though this option is enabled by default, you can fine tune the behavior by setting `packageRules` if you want to use it selectively (e.g. per-package).
+You can also fine-tune the behavior by setting `packageRules` if you want to use it selectively (e.g. per-package).
+
+Note that the outcome of `rebaseWhen=auto` can differ when `platformAutomerge=true` is enable.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1682]:
```suggestion
Note that the outcome of `rebaseWhen=auto` can differ when `platformAutomerge=true`.
```

Maybe this is cleaner now that the text has been changed.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Last change and then I'm done. 😄

**Review by HonkingGoose — DISMISSED** [review]

**Pull request merged**